File: frontend/src/utils/permissions.ts
import { User } from '../lib/api';

type Role = 'Admin' | 'Leader' | 'Sales Rep' | 'Inventory Manager' | 'Warehouse Staff' | 'Sales';

export const hasRole = (user: User | null, role: Role): boolean => {
    return user?.role === role;
};

export const isAdmin = (user: User | null): boolean => {
    return hasRole(user, 'Admin');
};

export const isLeader = (user: User | null): boolean => {
    return hasRole(user, 'Leader');
};

export const canManageCustomers = (user: User | null): boolean => {
    return hasRole(user, 'Admin') || hasRole(user, 'Leader') || hasRole(user, 'Inventory Manager');
};

export const canManageBrands = (user: User | null): boolean => {
    return hasRole(user, 'Inventory Manager') || hasRole(user, 'Admin') || hasRole(user, 'Leader');
};

export const canManageItems = (user: User | null): boolean => {
    return hasRole(user, 'Inventory Manager') || hasRole(user, 'Warehouse Staff') || hasRole(user, 'Admin') || hasRole(user, 'Leader');
};

export const canManageTransactions = (user: User | null): boolean => {
    return hasRole(user, 'Inventory Manager') || hasRole(user, 'Warehouse Staff') || hasRole(user, 'Admin') || hasRole(user, 'Leader');
};

export const canApproveTransactions = (user: User | null): boolean => {
    return hasRole(user, 'Inventory Manager') || hasRole(user, 'Warehouse Staff') || hasRole(user, 'Admin') || hasRole(user, 'Leader');
};

export const canViewAlerts = (user: User | null): boolean => {
    return hasRole(user, 'Inventory Manager') || hasRole(user, 'Warehouse Staff') || hasRole(user, 'Admin') || hasRole(user, 'Leader');
};

export const canExportData = (_: User | null): boolean => {
    return true; // All roles can export data
};

export const canViewDashboard = (_: User | null): boolean => {
    return true; // All roles can view dashboard
};

export const canReserveGoods = (_: User | null): boolean => {
    return true; // All roles can reserve goods
};

export const canViewOwnTransactions = (_: User | null): boolean => {
    return true; // All roles can view their own transactions
};

export const isSales = (user: User | null): boolean => {
    return hasRole(user, 'Sales') || hasRole(user, 'Sales Rep');
};

export const isWarehouseStaff = (user: User | null): boolean => {
    return hasRole(user, 'Warehouse Staff');
};

export const isInventoryManager = (user: User | null): boolean => {
    return hasRole(user, 'Inventory Manager') || hasRole(user, 'Admin') || hasRole(user, 'Leader');
};

export const canCancelOwnPendingTransaction = (user: User | null, transaction: any): boolean => {
    if (!user || !transaction) return false;
    if (isInventoryManager(user) || isWarehouseStaff(user) || isAdmin(user) || isLeader(user)) return true;
    if (isSales(user) && transaction.transaction_status === 'Pending') return true;
    return false;
};

File: backend/inventory/permissions.py
from rest_framework import permissions

class IsInventoryManager(permissions.BasePermission):
    def has_permission(self, request, view):
        return request.user.role == 'Inventory Manager'

class IsWarehouseStaff(permissions.BasePermission):
    def has_permission(self, request, view):
        return request.user.role == 'Warehouse Staff'

class IsSales(permissions.BasePermission):
    def has_permission(self, request, view):
        return request.user.role == 'Sales'

class CanManageBrands(permissions.BasePermission):
    def has_permission(self, request, view):
        return request.user.role in ['Inventory Manager']

class CanManageItems(permissions.BasePermission):
    def has_permission(self, request, view):
        return request.user.role in ['Inventory Manager', 'Warehouse Staff']

class CanManageTransactions(permissions.BasePermission):
    def has_permission(self, request, view):
        return request.user.role in ['Inventory Manager', 'Warehouse Staff']

class CanApproveTransactions(permissions.BasePermission):
    def has_permission(self, request, view):
        return request.user.role in ['Inventory Manager', 'Warehouse Staff']

class CanViewAlerts(permissions.BasePermission):
    def has_permission(self, request, view):
        return request.user.role in ['Inventory Manager', 'Warehouse Staff']

class CanExportData(permissions.BasePermission):
    def has_permission(self, request, view):
        return True  # All roles can export data

class CanViewDashboard(permissions.BasePermission):
    def has_permission(self, request, view):
        return True  # All roles can view dashboard

class CanReserveGoods(permissions.BasePermission):
    def has_permission(self, request, view):
        return request.user.role in ['Inventory Manager', 'Warehouse Staff', 'Sales']

class CanViewOwnTransactions(permissions.BasePermission):
    def has_permission(self, request, view):
        return True  # All roles can view their own transactions

    def has_object_permission(self, request, view, obj):
        return obj.account == request.user 

class ReadOnlyOrCanManageItems(permissions.BasePermission):
    """
    Allow all authenticated users to read, but only managers/staff to write.
    """
    def has_permission(self, request, view):
        if request.method in permissions.SAFE_METHODS:
            return request.user and request.user.is_authenticated
        return request.user.role in ['Inventory Manager', 'Warehouse Staff']

class ReadOnlyOrCanManageBrands(permissions.BasePermission):
    """
    Allow all authenticated users to read, but only managers to write.
    """
    def has_permission(self, request, view):
        if request.method in permissions.SAFE_METHODS:
            return request.user and request.user.is_authenticated
        return request.user.role == 'Inventory Manager' 

class ReadOnlyOrCanManageTransactions(permissions.BasePermission):
    """
    Allow all authenticated users to read, but only managers/staff to write.
    """
    def has_permission(self, request, view):
        if request.method in permissions.SAFE_METHODS:
            return request.user and request.user.is_authenticated
        return request.user.role in ['Inventory Manager', 'Warehouse Staff'] 

class CanCancelOwnPendingTransaction(permissions.BasePermission):
    def has_object_permission(self, request, view, obj):
        print('DEBUG: user', request.user)
        print('DEBUG: user.role', getattr(request.user, 'role', None))
        print('DEBUG: obj.transaction_status', getattr(obj, 'transaction_status', None))
        if request.user.role in ['Inventory Manager', 'Warehouse Staff']:
            return True
        if request.user.role == 'Sales' and obj.transaction_status == 'Pending':
            return True
        return False